### PR TITLE
[NVIDIA] Log FW Update error

### PR DESCRIPTION
**Pull request by DavidZagury**
#### Why I did it
To have information about the cause of FW update failure in the log.

#### How I did it
In case the FW update has failed the update tool will give the cause of the failure in the output in the last line, starting with "Fail".
When running the tool, in case of a failed update, we will parse the output to retrieve the cause and log it.
```
Device #1:
 ----------
 
 Device Type:      ConnectX6DX
   Part Number:      MCX623106AN-CDA_Ax
   Description:      ConnectX-6 Dx EN adapter card; 100GbE; Dual-port QSFP56; PCIe 4.0/3.0 x16;
   PSID:             MT_0000000359
   PCI Device Name:  /dev/mst/mt4125_pciconf0
   Base GUID:        0c42a103007d22d4
   Base MAC:         0c42a17d22d4
   Versions:         Current        Available     
      FW             22.32.0498     22.32.0498    
      PXE            3.6.0500       3.6.0500      
      UEFI           14.25.0015     14.25.0015    
 
 Status:           Forced update required
 
---------
 Found 1 device(s) requiring firmware update...
 
Device #1: Updating FW ...     
 FSMST_INITIALIZE -   OK          
 Writing Boot image component -   OK          
 Fail : The Digest in the signature is wrong
```

#### How to verify it
Create a scenario where FW update fails (i.e. corrupted FW) and use the mlnx-fw-update script to try to update the FW.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211

### CODE REVIEW

**Review by keboliu — APPROVED** [review]

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 1]
```diff
@@ -186,6 +186,23 @@ function RunCmd() {
...
+    if [[ "${VERBOSE_LEVEL}" -eq "${VERBOSE_MAX}" ]]; then
+            output=$(eval "${COMMAND}")
+    else
+            output=$(eval "${COMMAND}") >/dev/null 2>&1
+    fi
+
+    ERROR_CODE="$?"
+    if [[ "${ERROR_CODE}" != "${EXIT_SUCCESS}" ]]; then
+        failure_msg=${output#*Fail : }
+        ExitFailure "FW Update command failed: ${COMMAND} with error:" ${failure_msg}
```
Comment (keboliu) [inline platform/mellanox/mlnx-fw-upgrade.j2:202]: ExitFailure "FW Update command : ${COMMAND} failed with error:" ${failure_msg}

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 2]
```diff
@@ -186,6 +186,23 @@ function RunCmd() {
...
+    if [[ "${VERBOSE_LEVEL}" -eq "${VERBOSE_MAX}" ]]; then
+            output=$(eval "${COMMAND}")
+    else
+            output=$(eval "${COMMAND}") >/dev/null 2>&1
+    fi
+
+    ERROR_CODE="$?"
+    if [[ "${ERROR_CODE}" != "${EXIT_SUCCESS}" ]]; then
+        failure_msg=${output#*Fail : }
+        ExitFailure "FW Update command: ${COMMAND} failed with error:" ${failure_msg}
```
Comment (nazariig) [inline platform/mellanox/mlnx-fw-upgrade.j2:202]: @DavidZagury it won't print to the syslog. Suggest to rework.

**Review by nazariig — CHANGES_REQUESTED** [review]

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 3]
```diff
@@ -186,6 +210,23 @@ function RunCmd() {
     fi
 }
 
+function RunFwUpdateCmd() {
+    local ERROR_CODE="${EXIT_SUCCESS}"
+    local COMMAND="${BURN_CMD} $@"
+
+    if [[ "${VERBOSE_LEVEL}" -eq "${VERBOSE_MAX}" ]]; then
+            output=$(eval "${COMMAND}")
```
Comment (nazariig) [inline platform/mellanox/mlnx-fw-upgrade.j2:218]: @DavidZagury please use tab with 4 spaces

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 4]
```diff
@@ -186,6 +210,23 @@ function RunCmd() {
...
+    else
+            output=$(eval "${COMMAND}") >/dev/null 2>&1
```
Comment (nazariig) [inline platform/mellanox/mlnx-fw-upgrade.j2:220]: @DavidZagury please use tab with 4 spaces

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 5]
```diff
@@ -186,6 +210,23 @@ function RunCmd() {
...
+    fi
+
+    ERROR_CODE="$?"
+    if [[ "${ERROR_CODE}" != "${EXIT_SUCCESS}" ]]; then
+        failure_msg=${output#*Fail : }
```
Comment (nazariig) [inline platform/mellanox/mlnx-fw-upgrade.j2:225]: @DavidZagury please wrap command with double quotes

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 6]
```diff
@@ -186,6 +210,23 @@ function RunCmd() {
...
+        ExitFailure "FW Update command: ${COMMAND} failed with error:" ${failure_msg}
```
Comment (nazariig) [inline platform/mellanox/mlnx-fw-upgrade.j2:226]: @DavidZagury please include `${failure_msg}` into the common double quotes

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 7]
```diff
@@ -186,6 +210,23 @@ function RunCmd() {
...
+    if [[ "${VERBOSE_LEVEL}" -eq "${VERBOSE_MAX}" ]]; then
+        output=$(eval "${COMMAND}")
+    else
+        output=$(eval "${COMMAND}") >/dev/null 2>&1
+    fi
+
+    ERROR_CODE="$?"
+    if [[ "${ERROR_CODE}" != "${EXIT_SUCCESS}" ]]; then
+        failure_msg="${output#*Fail : }"
+        ExitFailure "FW Update command: ${COMMAND} failed with error:" ${failure_msg}
```
Comment (nazariig) [inline platform/mellanox/mlnx-fw-upgrade.j2:226]:
```suggestion
        ExitFailure "FW Update command: ${COMMAND} failed with error: ${failure_msg}"
```

**Review by nazariig — CHANGES_REQUESTED** [review]

**Pull request closed without merging**